### ISSUE TITLE
Support for type hints

### ISSUE THREAD

**Issue by ben-z** [crdoconnor/strictyaml#90]
Feature request

Does strictyaml work well with [type hints](https://docs.python.org/3/library/typing.html)? I can't find any documentation for this. I would like to have this functionality so I won't need to define my types twice.

**Comment by crdoconnor** [comment]
No, not yet, but I'm open to the idea. Can you share some example code that demonstrates the problem?

**Comment by ben-z** [comment]
My use case is this:

```python
from strictyaml import load as load_yaml, Map

schema = MAP({ ... })

class SomeClass():
  config: schema # I want this part
  def __init__(self, config: schema):
    self.config = config
  
  @static_method
  def from_yaml(yaml_str: str):
    return SomeClass(load_yaml(yaml_str, schema))

if __name__ == "__main":
  ...
  obj = SomeClass.from_yaml(yaml_str)
```

Does the above look possible?

**Comment by opk12** [comment]
An example the other way around: the program wants a list of TypedDict from a yaml,
`strictyaml.load(yaml, List[Movie])` would auto-generate a schema.

Based on the [PEP 589](https://www.python.org/dev/peps/pep-0589/) (TypeDict) Movies snippet.

```
from typing import cast, List, Optional, TypedDict

import strictyaml
from strictyaml import Int, Map, Seq, Str

class Location(TypedDict):
    country: str

SCHEMA_LOCATION = Map(
    {
        "country": Str()
    }
)

class Movie(TypedDict):
    name: str
    year: int
    locations: List[Location]
    comments: Optional[List[str]]

SCHEMA_MOVIE = Map(
    {
        "name": Str(),
        "year": Int(),
        "locations": Seq(SCHEMA_LOCATION),
        strictyaml.Optional("comments", default=None): Seq(Str()),
    }
)

def movies(yaml: str) -> List[Movie]:
    # current, working code:
    data = strictyaml.load(yaml, schema=Seq(SCHEMA_MOVIE)).data
    return cast(List[Movie], data)

    # ideally it would be:
    return strictyaml.load(yaml, hint=List[Movie])
```

These are minor issues, but it would be nice if I could avoid
1. type information duplication
2. `cast()`
3. looking up the docs for the `strictyaml` equivalents of `typing`'s `Union`, etc

`strictyaml.Optional` is qualified above, for the collision with `typing.Optional`.

In this example, `load()` would detect a `List` [1] of something that has `__annotations__` [2].

[1] `List.__origin__ == list` should work, according to the `typing` module source code, but maybe a better way is documented somewhere
[2] per [PEP 3107 § Accessing Function Annotations](https://www.python.org/dev/peps/pep-3107/#accessing-function-annotations)

**Comment by opk12** [comment]
It seems the above `movies()` cast is incorrect, because `OrderedDict`s are currently generated, per issue #80, while `TypedDict` is basically a constructor/type hint for a plain `dict`. This is yet another reason to avoid `cast()`.

**Comment by opk12** [comment]
For completeness and myself: the rationale for the cast above is that `.data` [contains](https://hitchdev.com/strictyaml/using/alpha/compound/mapping-yaml-object/) `just other dicts, lists and strings/ints/etc`, as opposed to custom containers, and `TypedDict` is essentially a constructor and type hint for a plain `dict` (PEP: `the runtime type of a TypedDict object will always be just dict (it is never a subclass of dict)`).

**Comment by RichardFevrier** [comment]
👀
